feat: add opentelemetry plugin (#6119)
apache/apisix: docs/en/latest/plugins/opentelemetry.md

| ------------ | ------ | ------ | -------- | ------------ | ----------------------------------------------------- |
| sampler | object | optional | | | sampling config
| sampler.name | string | optional | always_off | ["always_on", "always_off", "trace_id_ratio", "parent_base"] | sampling strategy，always_on：sampling all；always_off：sampling nothing；trace_id_ratio：base trace id percentage；parent_base：use parent decision, otherwise determined by root
| sampler.options | object | optional | | {fraction = 0, root = {name = "always_off"}} | sampling strategy parameters
| sampler.options.fraction | number | optional | 0 | [0, 1] | trace_id_ratio fraction
| sampler.options.root | object | optional | {name = "always_off", options = {fraction = 0}} | | parent_base root sampler
| sampler.options.root.name | string | optional | always_off | ["always_on", "always_off", "trace_id_ratio"] | sampling strategy
| sampler.options.root.options | object | optional | {fraction = 0} | | sampling strategy parameters
| sampler.options.root.options.fraction | number | optional | 0 | [0, 1] | trace_id_ratio fraction

> | additional_attributes | array[string] | optional | | | append to trace span attributes

Reviewer: ```suggestion
| additional_attributes | array[string] | optional | | | attributes (variable and its value) which will be appended to the trace span 
```

> | additional_attributes[0] | string | required | | | key of ctx.var

Reviewer: ```suggestion
| additional_attributes[0] | string | required | | | APISIX or Nginx variable, like `http_header` or `route_id`
```

## How To Enable

First of all, enable the opentelemetry plugin in the `config.yaml`:

```yaml
# Add this in config.yaml
plugins:
  - ... # plugin you need
  - opentelemetry
```

Then reload APISIX.

Here's an example, enable the opentelemetry plugin on the specified route:

```shell
curl http://127.0.0.1:9080/apisix/admin/routes/1  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "methods": ["GET"],
    "uris": [
        "/uid/*"
    ],
    "plugins": {
        "opentelemetry": {
            sampler": {
                "name": "always_on",
            }
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "10.110.149.175:8089": 1
        }
    }
}'
```

## How to set collecting

> We can set the collecting by specifying the configuration in `conf/config.yaml`.

Reviewer 2: ```suggestion
You can set the collecting by specifying the configuration in `conf/config.yaml`.
```
Based on the pronouns used throughout the document : )

| Name         | Type   | Default  | Description                                                          |
| ------------ | ------ | -------- | ----------------------------------------------------- |

> | trace_id_source | enum | random | alternate use x-request-id as trace id, valid value is `random` or `x-request-id`, if use `x-request-id`, please make sure it match regex pattern `[0-9a-f]{32}` |

Reviewer: ```suggestion
| trace_id_source | enum | random | the source of trace id, the valid value is `random` or `x-request-id`. If `x-request-id` is set, the value of `x-request-id` request header will be used as trace id. Please make sure it match regex pattern `[0-9a-f]{32}` |
```

---
title: opentelemetry
---

<!--
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

## Summary

- [**Name**](#name)
- [**Attributes**](#attributes)
- [**How To Enable**](#how-to-enable)
- [**How to set collecting**](#how-to-set-collecting)
- [**Disable Plugin**](#disable-plugin)

## Name

> opentelemetry report Tracing data according to [opentelemetry specification](https://github.com/open-telemetry/opentelemetry-specification).

Reviewer 3: ```suggestion
[OpenTelemetry](https://opentelemetry.io/) report Tracing data according to [opentelemetry specification](https://github.com/open-telemetry/opentelemetry-specification).
```

> Just support reporting in HTTP with Content-Type=application/x-protobuf, the specification: [OTLP/HTTP Request](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp-request)。

Reviewer 3: ```suggestion
Just support reporting in `HTTP` with `Content-Type=application/x-protobuf`, the specification: [OTLP/HTTP Request](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp-request)。
```

| sampler.options | object | optional | | {fraction = 0, root = {name = "always_off"}} | sampling strategy parameters
| sampler.options.fraction | number | optional | 0 | [0, 1] | trace_id_ratio fraction
| sampler.options.root | object | optional | {name = "always_off", options = {fraction = 0}} | | parent_base root sampler
| sampler.options.root.name | string | optional | always_off | ["always_on", "always_off", "trace_id_ratio"] | sampling strategy
| sampler.options.root.options | object | optional | {fraction = 0} | | sampling strategy parameters
| sampler.options.root.options.fraction | number | optional | 0 | [0, 1] | trace_id_ratio fraction
| tags | array[object] | optional | | | append to trace span attributes
| tags.position | string | required | | ["http", "arg", "cookie"] | where variable in
| tags.name | string | required | | | variable name

## How To Enable

First of all, enable the opentelemetry plugin in the `config.yaml`:

> ```

Reviewer 3: ```suggestion
```yaml
```

# Add this in config.yaml
plugins:
  - ... # plugin you need

>   - opentelemetry

Reviewer: Let's add a commented placeholder like
https://github.com/apache/apisix/blob/94088cde46a0c12149ba8dcfeb7bf4bfdb1ff869/conf/config-default.yaml#L384
